// stock-snapshot/src/types/StockStatistics.ts
export interface StockStatistics {
  marketCap: number | null;
  enterpriseValue: number | null;
  earningsDate: string | null;
  exDividendDate: string | null;
  sharesOutstanding: number | null;
  sharesChangeYoY: number | null;
  sharesChangeQoQ: number | null;
  ownedByInstitutions: number | null;
  peRatio: number | null;
  forwardPERatio: number | null;
  psRatio: number | null;
  pbRatio: number | null;
  ptbvRatio: number | null;
  pfcfRatio: number | null;
  pocfRatio: number | null;
  pegRatio: number | null;
  evEarnings: number | null;
  evSales: number | null;
  evEbitda: number | null;
  evEbit: number | null;
  evFcf: number | null;
  currentRatio: number | null;
  quickRatio: number | null;
  debtToEquity: number | null;
  debtToEbitda: number | null;
  debtToFcf: number | null;
  interestCoverage: number | null;
  returnOnEquity: number | null;
  returnOnAssets: number | null;
  returnOnInvestedCapital: number | null;
  returnOnCapitalEmployed: number | null;
  beta5Y: number | null;
  priceChange52W: number | null;
  movingAverage50D: number | null;
  movingAverage200D: number | null;
  rsi: number | null;
  averageVolume20D: number | null;
  revenue: number | null;
  grossProfit: number | null;
  operatingIncome: number | null;
  pretaxIncome: number | null;
  netIncome: number | null;
  ebitda: number | null;
  ebit: number | null;
  eps: number | null;
  cash: number | null;
  totalDebt: number | null;
  netCash: number | null;
  netCashPerShare: number | null;
  bookValue: number | null;
  bookValuePerShare: number | null;
  workingCapital: number | null;
  operatingCashFlow: number | null;
  capitalExpenditures: number | null;
  freeCashFlow: number | null;
  freeCashFlowPerShare: number | null;
  grossMargin: number | null;
  operatingMargin: number | null;
  pretaxMargin: number | null;
  profitMargin: number | null;
  ebitdaMargin: number | null;
  ebitMargin: number | null;
  fcfMargin: number | null;
  dividendPerShare: number | null;
  dividendYield: number | null;
  dividendGrowth: number | null;
  payoutRatio: number | null;
  buybackYield: number | null;
  shareholderYield: number | null;
  earningsYield: number | null;
  fcfYield: number | null;
  altmanZScore: number | null;
  piotroskiFScore: number | null;
}
