// stock-snapshot/dist/types/balanceSheet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BalanceSheetField = void 0;
// รายชื่อฟิลด์ของ Balance Sheet
var BalanceSheetField;
(function (BalanceSheetField) {
    BalanceSheetField["CashAndEquivalents"] = "Cash & Equivalents";
    BalanceSheetField["ShortTermInvestments"] = "Short-Term Investments";
    BalanceSheetField["CashAndShortTermInvestments"] = "Cash & Short-Term Investments";
    BalanceSheetField["CashGrowth"] = "Cash Growth";
    BalanceSheetField["AccountsReceivable"] = "Accounts Receivable";
    BalanceSheetField["OtherReceivables"] = "Other Receivables";
    BalanceSheetField["Receivables"] = "Receivables";
    BalanceSheetField["Inventory"] = "Inventory";
    BalanceSheetField["PrepaidExpenses"] = "Prepaid Expenses";
    BalanceSheetField["OtherCurrentAssets"] = "Other Current Assets";
    BalanceSheetField["TotalCurrentAssets"] = "Total Current Assets";
    BalanceSheetField["PropertyPlantEquipment"] = "Property, Plant & Equipment";
    BalanceSheetField["LongTermInvestments"] = "Long-Term Investments";
    BalanceSheetField["OtherIntangibleAssets"] = "Other Intangible Assets";
    BalanceSheetField["LongTermAccountsReceivable"] = "Long-Term Accounts Receivable";
    BalanceSheetField["LongTermDeferredTaxAssets"] = "Long-Term Deferred Tax Assets";
    BalanceSheetField["OtherLongTermAssets"] = "Other Long-Term Assets";
    BalanceSheetField["TotalAssets"] = "Total Assets";
    BalanceSheetField["AccountsPayable"] = "Accounts Payable";
    BalanceSheetField["AccruedExpenses"] = "Accrued Expenses";
    BalanceSheetField["ShortTermDebt"] = "Short-Term Debt";
    BalanceSheetField["CurrentPortionOfLongTermDebt"] = "Current Portion of Long-Term Debt";
    BalanceSheetField["CurrentPortionOfLeases"] = "Current Portion of Leases";
    BalanceSheetField["CurrentIncomeTaxesPayable"] = "Current Income Taxes Payable";
    BalanceSheetField["CurrentUnearnedRevenue"] = "Current Unearned Revenue";
    BalanceSheetField["OtherCurrentLiabilities"] = "Other Current Liabilities";
    BalanceSheetField["TotalCurrentLiabilities"] = "Total Current Liabilities";
    BalanceSheetField["LongTermDebt"] = "Long-Term Debt";
    BalanceSheetField["LongTermLeases"] = "Long-Term Leases";
    BalanceSheetField["LongTermUnearnedRevenue"] = "Long-Term Unearned Revenue";
    BalanceSheetField["OtherLongTermLiabilities"] = "Other Long-Term Liabilities";
    BalanceSheetField["TotalLiabilities"] = "Total Liabilities";
    BalanceSheetField["CommonStock"] = "Common Stock";
    BalanceSheetField["AdditionalPaidInCapital"] = "Additional Paid-In Capital";
    BalanceSheetField["RetainedEarnings"] = "Retained Earnings";
    BalanceSheetField["ComprehensiveIncomeAndOther"] = "Comprehensive Income & Other";
    BalanceSheetField["TotalCommonEquity"] = "Total Common Equity";
    BalanceSheetField["MinorityInterest"] = "Minority Interest";
    BalanceSheetField["ShareholdersEquity"] = "Shareholders' Equity";
    BalanceSheetField["TotalLiabilitiesAndEquity"] = "Total Liabilities & Equity";
    BalanceSheetField["TotalDebt"] = "Total Debt";
    BalanceSheetField["NetCashDebt"] = "Net Cash (Debt)";
    BalanceSheetField["NetCashGrowth"] = "Net Cash Growth";
    BalanceSheetField["NetCashPerShare"] = "Net Cash Per Share";
    BalanceSheetField["FilingDateSharesOutstanding"] = "Filing Date Shares Outstanding";
    BalanceSheetField["TotalCommonSharesOutstanding"] = "Total Common Shares Outstanding";
    BalanceSheetField["WorkingCapital"] = "Working Capital";
    BalanceSheetField["BookValuePerShare"] = "Book Value Per Share";
    BalanceSheetField["TangibleBookValue"] = "Tangible Book Value";
    BalanceSheetField["TangibleBookValuePerShare"] = "Tangible Book Value Per Share";
    BalanceSheetField["Land"] = "Land";
    BalanceSheetField["Buildings"] = "Buildings";
    BalanceSheetField["Machinery"] = "Machinery";
    BalanceSheetField["ConstructionInProgress"] = "Construction In Progress";
})(BalanceSheetField || (exports.BalanceSheetField = BalanceSheetField = {}));
