// stock-snapshot/dist/types/financial.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FinancialField = void 0;
// รายชื่อฟิลด์ของ Income Statement
var FinancialField;
(function (FinancialField) {
    FinancialField["OperatingRevenue"] = "Operating Revenue";
    FinancialField["OtherRevenue"] = "Other Revenue";
    FinancialField["Revenue"] = "Revenue";
    FinancialField["RevenueGrowth"] = "Revenue Growth (YoY)";
    FinancialField["CostOfRevenue"] = "Cost of Revenue";
    FinancialField["GrossProfit"] = "Gross Profit";
    FinancialField["SellingGeneralAdmin"] = "Selling, General & Admin";
    FinancialField["OtherOperatingExpenses"] = "Other Operating Expenses";
    FinancialField["OperatingExpenses"] = "Operating Expenses";
    FinancialField["OperatingIncome"] = "Operating Income";
    FinancialField["InterestExpense"] = "Interest Expense";
    FinancialField["InterestInvestmentIncome"] = "Interest & Investment Income";
    FinancialField["EarningsFromEquityInvestments"] = "Earnings From Equity Investments";
    FinancialField["CurrencyExchangeGainLoss"] = "Currency Exchange Gain (Loss)";
    FinancialField["OtherNonOperatingIncomeExpenses"] = "Other Non Operating Income (Expenses)";
    FinancialField["EBTExcludingUnusualItems"] = "EBT Excluding Unusual Items";
    FinancialField["GainLossOnSaleOfAssets"] = "Gain (Loss) on Sale of Assets";
    FinancialField["AssetWritedown"] = "Asset Writedown";
    FinancialField["PretaxIncome"] = "Pretax Income";
    FinancialField["IncomeTaxExpense"] = "Income Tax Expense";
    FinancialField["EarningsFromContinuingOperations"] = "Earnings From Continuing Operations";
    FinancialField["MinorityInterestInEarnings"] = "Minority Interest in Earnings";
    FinancialField["NetIncome"] = "Net Income";
    FinancialField["NetIncomeToCommon"] = "Net Income to Common";
    FinancialField["NetIncomeGrowth"] = "Net Income Growth";
    FinancialField["SharesOutstandingBasic"] = "Shares Outstanding (Basic)";
    FinancialField["SharesOutstandingDiluted"] = "Shares Outstanding (Diluted)";
    FinancialField["EPSBasic"] = "EPS (Basic)";
    FinancialField["EPSDiluted"] = "EPS (Diluted)";
    FinancialField["EPSGrowth"] = "EPS Growth";
    FinancialField["FreeCashFlow"] = "Free Cash Flow";
    FinancialField["FreeCashFlowPerShare"] = "Free Cash Flow Per Share";
    FinancialField["DividendPerShare"] = "Dividend Per Share";
    FinancialField["DividendGrowth"] = "Dividend Growth";
    FinancialField["GrossMargin"] = "Gross Margin";
    FinancialField["OperatingMargin"] = "Operating Margin";
    FinancialField["ProfitMargin"] = "Profit Margin";
    FinancialField["FreeCashFlowMargin"] = "Free Cash Flow Margin";
    FinancialField["EBITDA"] = "EBITDA";
    FinancialField["EBITDAMargin"] = "EBITDA Margin";
    FinancialField["DAForEBITDA"] = "D&A For EBITDA";
    FinancialField["EBIT"] = "EBIT";
    FinancialField["EBITMargin"] = "EBIT Margin";
    FinancialField["EffectiveTaxRate"] = "Effective Tax Rate";
    FinancialField["RevenueAsReported"] = "Revenue as Reported";
})(FinancialField || (exports.FinancialField = FinancialField = {}));
